huggan/tensorflow/dcgan/train.py

```diff
@@ -0,0 +1,213 @@
+    parser.add_argument("--number_of_examples_to_generate", type=int, default=4, help="Number of examples to be generated in inference mode")
+    parser.add_argument(
+        "--generator_hidden_size",
+        type=int,
+        default=28,
+        help="Hidden size of the generator's feature maps.",
+    )
+    parser.add_argument("--latent_dim", type=int, default=100, help="Dimensionality of the latent space.")
+
+    parser.add_argument(
+        "--discriminator_hidden_size",
+        type=int,
+        default=28,
+        help="Hidden size of the discriminator's feature maps.",
+    )
+    parser.add_argument(
+        "--image_size",
+        type=int,
+        default=28,
+        help="Spatial size to use when resizing images for training.",
+    )
+    parser.add_argument(
+        "--num_channels",
+        type=int,
+        default=3,
+        help="Number of channels in the training images. For color images this is 3.",
+    )
+    parser.add_argument("--num_epochs", type=int, default=5, help="number of epochs of training")
+    parser.add_argument("--lr", type=float, default=0.01, help="Learning rate")
+    parser.add_argument("--output_dir", type=Path, default=Path("./output"), help="Name of the directory to dump generated images during training.")
+    parser.add_argument(
+        "--push_to_hub",
+        action="store_true",
+        help="Whether to push the model to the HuggingFace hub after training.",
+        )
+    parser.add_argument(
+        "--model_name",
+        default=None,
+        type=str,
+        help="Name of the model on the hub.",
+    )
+    parser.add_argument(
+        "--organization_name",
+        default="huggan",
+        type=str,
+        help="Organization name to push to, in case args.push_to_hub is specified.",
+    )
+    args = parser.parse_args()
+    
+    if args.push_to_hub:
+        assert args.output_dir is not None, "Need an `output_dir` to create a repo when `--push_to_hub` is passed."
+        assert args.model_name is not None, "Need a `model_name` to create a repo when `--push_to_hub` is passed."
+
+    if args.output_dir is not None:
+        os.makedirs(args.output_dir, exist_ok=True)
+    
+    return args
+
+
+
+
+def preprocess_dataset(dataset, BATCH_SIZE, image_size):
+    preprocessed_images = []
+    print("Preprocessing dataset..")
+    for example in range(len(dataset["train"])):
+        img = dataset["train"][example]["image"].convert('RGB') 
+        img = np.asarray(img) 
+        img = tf.keras.utils.img_to_array(PIL.Image.fromarray(img))
+        preprocessed_images.append(img)
+    train_images = np.asarray(preprocessed_images)
+    channel = train_images[0].shape[-1]
+    train_images = train_images.reshape(train_images.shape[0], image_size, image_size, channel).astype('float32')
+
+    train_images = (train_images - 127.5) / 127.5 
+    train_dataset = tf.data.Dataset.from_tensor_slices(train_images).batch(BATCH_SIZE)
+    return train_dataset, channel
```

Comment:
Hmm not sure if it should be processed like that. The big features of Datasets are .map, .with_transform and .set_transform (and optionally set_format).

You shouldn't manually iterate over the rows of the dataset, the aformentioned functions do that for you (see also my Pytorch script: the function you define should just be defined for a batch of examples)

Reply:
I had a problem in process and wanted to get script out ASAP. I can do this if you want that.

Reply:
I did this instead, `PrefetchDataset` returned by `to_tf_dataset` doesn't work well with `model()` and dataset map was returning elements like [[image],[another image]] so I unpacked and made another list that `from_tensor_slices` wanted.
```
def preprocess_images(dataset, args):
    processed_dataset = dataset["test"].map(preprocess)
    new_dataset = []
    
    for image in processed_dataset["pixel_values"]:
        new_dataset.append(image[0])
    
    new_dataset = tf.data.Dataset.from_tensor_slices(new_dataset).batch(args.batch_size)
    return new_dataset
```

Reply:
cc'ing @Rocketknight1 here, apparently` .to_tf_dataset` wasn't ideal for vision datasets

Reply:
added `to_tf_dataset` as well

Reply:
`def preprocess(examples):
    images = (np.asarray(examples["image"]).astype('float32')- 127.5) / 127.5
    images = np.expand_dims(images, -1)
    examples["pixel_values"] = images
    return examples



def preprocess_images(dataset, args):
    
 
    data_collator = DefaultDataCollator(return_tensors="tf")


    processed_dataset = dataset.map(preprocess)
    
    tf_train_dataset = processed_dataset.to_tf_dataset(
	    columns=['pixel_values'],
	    shuffle=True,
	    batch_size=args.batch_size,
	    collate_fn=data_collator)

    return tf_train_dataset`

I didn't expand dimensions previously that's why it wasn't working